### PR TITLE
Array speedup

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

File: src/solvers/flattening/arrays.cpp [hunk 1]
```diff
@@ -8,6 +8,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "arrays.h"
 
+#include <cassert>
+#include <iostream>
+
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: Don't include these. All outputs should be in appropriate mstreamt and assertions should be replaced by invariants and such.

File: src/solvers/flattening/arrays.cpp [hunk 2]
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
   // check types
   if(!base_type_eq(op0.type(), op1.type(), ns))
   {
-    prop.error() << equality.pretty() << messaget::eom;
-    DATA_INVARIANT(
-      false,
-      "record_array_equality got equality without matching types");
+    std::cout << equality.pretty() << '\n';
+    throw "record_array_equality got equality without matching types";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: use INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 3]
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
...
   }
 
-  DATA_INVARIANT(
-    ns.follow(op0.type()).id()==ID_array,
-    "record_array_equality parameter should be array-typed");
+  INVARIANT(ns.follow(op0.type()).id()==ID_array,
+            "equality has array type");
+
+  std::size_t a1=arrays.number(op0);
+  std::size_t a2=arrays.number(op1);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: const ?

File: src/solvers/flattening/arrays.cpp [hunk 4]
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
...
+  return l;
 }
 
 void arrayst::collect_indices(const exprt &expr)
 {
   if(expr.id()!=ID_index)
   {
-    forall_operands(op, expr) collect_indices(*op);
+    forall_operands(op, expr)
+      collect_indices(*op);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: I would suggest `std::for_each(expr.operands().begin(), expr.operands().end(), collect_indices)`

File: src/solvers/flattening/arrays.cpp [hunk 5]
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
   const array_typet &array_type=
     to_array_type(ns.follow(a.type()));
 
+  // remember it
+  wegt::node_indext a_ind=arrays.number(a);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: `a_index` (and maybe const)

File: src/solvers/flattening/arrays.cpp [hunk 6]
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
...
   {
     const with_exprt &with_expr=to_with_expr(a);
 
     // check types
     if(!base_type_eq(array_type, with_expr.old().type(), ns))
     {
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got 'with' without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got 'with' without matching types";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 7]
```diff
@@ -99,9 +85,7 @@ void arrayst::collect_indices(const exprt &expr)
         to_array_type(array_op_type);
 
       if(is_unbounded_array(array_type))
-      {
         record_array_index(e);
-      }
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp:122]: Could be part of a separate reformatting commit.

File: src/solvers/flattening/arrays.cpp [hunk 8]
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
...
+      throw "collect_arrays got 'with' without matching types";
     }
 
-    arrays.make_union(a, with_expr.old());
-    collect_arrays(with_expr.old());
+    collect_indices(with_expr.where());
+    collect_indices(with_expr.new_value());
+
+    // one undirected edge for each array update
+    std::size_t expr_old_ind=arrays.number(with_expr.old());
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: `expr_old_index` (const?)

File: src/solvers/flattening/arrays.cpp [hunk 9]
```diff
@@ -136,20 +126,14 @@ void arrayst::collect_arrays(const exprt &a)
     // check types
     if(!base_type_eq(array_type, update_expr.old().type(), ns))
     {
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(
-        false,
-        "collect_arrays got 'update' without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got 'update' without matching types";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 10]
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
     // check types
     if(!base_type_eq(array_type, if_expr.true_case().type(), ns))
     {
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got if without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got if without matching types";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 11]
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
...
     }
 
     // check types
     if(!base_type_eq(array_type, if_expr.false_case().type(), ns))
     {
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got if without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got if without matching types";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 12]
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
...
     }
 
-    arrays.make_union(a, if_expr.true_case());
-    arrays.make_union(a, if_expr.false_case());
+    // do weg edges
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp:171]: This comment does not add information. And what does weg mean?

File: src/solvers/flattening/arrays.cpp [hunk 13]
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
...
+    std::size_t expr_true_ind=arrays.number(if_expr.true_case());
+    std::size_t expr_false_ind=arrays.number(if_expr.false_case());
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: const, index

File: src/solvers/flattening/arrays.cpp [hunk 14]
```diff
@@ -182,10 +171,9 @@ void arrayst::collect_arrays(const exprt &a)
   }
   else if(a.id()==ID_member)
   {
-    DATA_INVARIANT(
-      to_member_expr(a).struct_op().id()==ID_symbol,
-      ("unexpected array expression: member with `"+
-       a.op0().id_string()+"'").c_str());
+    if(to_member_expr(a).struct_op().id()!=ID_symbol)
+      throw
+        "unexpected array expression: member with `"+a.op0().id_string()+"'";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: keep the DATA_INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 15]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
   else if(a.id()==ID_byte_update_little_endian ||
           a.id()==ID_byte_update_big_endian)
   {
-    DATA_INVARIANT(
-      false,
-      "byte_update should be removed before collect_arrays");
+    assert(false);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: don't use assert

File: src/solvers/flattening/arrays.cpp [hunk 16]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+    const exprt &op=to_typecast_expr(a).op();
 
-    arrays.make_union(a, a.op0());
-    collect_arrays(a.op0());
+    if(op.type().id()==ID_array)
+    {
+      collect_arrays(op);
+    }
+    else
+      throw "unexpected array type cast from "+op.type().id_string();
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: INVARIANT

File: src/solvers/flattening/arrays.cpp [hunk 17]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
   else if(a.id()==ID_index)
   {
-    // nested unbounded arrays
-    arrays.make_union(a, a.op0());
-    collect_arrays(a.op0());
+    // an array of arrays!
+    collect_arrays(to_index_expr(a).array());
   }
   else
+    throw "unexpected array expression (collect_arrays): `"+a.id_string()+"'";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: UNREACHABLE

File: src/solvers/flattening/arrays.h [hunk 18]
```diff
@@ -48,81 +49,97 @@ class arrayst:public equalityt
...
-    ARRAY_TYPECAST
-  };
+  virtual bool is_unbounded_array(const typet &type) const=0;
+    // (maybe this function should be partially moved here from boolbv)
 
-  struct lazy_constraintt
-  {
-    lazy_typet type;
-    exprt lazy;
+  bool must_be_different(const exprt &, const exprt &);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.h]: const?

File: src/solvers/flattening/arrays.cpp [hunk 19]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+}
+
+bool arrayst::must_be_different(
+  const exprt &x, const exprt &y)
+{
+  return false;
+}
+
+void arrayst::weg_path_condition(
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp:234]: should be documented

Comment (romainbrenguier) [reply src/solvers/flattening/arrays.cpp:234]: using a verb in the function name is generally clearer

File: src/solvers/flattening/arrays.cpp [hunk 20]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+{
+  for(const auto &pn : path)
   {
-    DATA_INVARIANT(
-      false,
-      ("unexpected array expression (collect_arrays): `"+
-       a.id_string()+"'").c_str());
+    std::cout << "edge: " << pn.n << "->" <<
+      pn.edge->first << " " <<
+      from_expr(arrays[pn.n]) << "\n";
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: don't write on `std::cout`

File: src/solvers/flattening/arrays.cpp [hunk 21]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+
+    const weg_edget &weg_edge=pn.edge->second;
+
+    if(weg_edge.update.is_not_nil())
+    {
+      const auto &with_expr=to_with_expr(weg_edge.update);
+      notequal_exprt inequality(with_expr.where(), index_a);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: const

File: src/solvers/flattening/arrays.cpp [hunk 22]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
 }
 
-/// adds array constraints (refine=true...lazily for the refinement loop)
-void arrayst::add_array_constraint(const lazy_constraintt &lazy, bool refine)
+void arrayst::process_weg_path(const weg_patht &path)
 {
-  if(lazy_arrays && refine)
+  INVARIANT(!path.empty(), "path is not empty");
+  wegt::node_indext a=path.front().n;
+  wegt::node_indext b=path.back().n;
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: const ?

File: src/solvers/flattening/arrays.cpp [hunk 23]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+      cond.push_back(literal_exprt(weg_edge.condition));
   }
 }
 
-/// adds array constraints (refine=true...lazily for the refinement loop)
-void arrayst::add_array_constraint(const lazy_constraintt &lazy, bool refine)
+void arrayst::process_weg_path(const weg_patht &path)
 {
-  if(lazy_arrays && refine)
+  INVARIANT(!path.empty(), "path is not empty");
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp:276]: PRECONDITION ?

File: src/solvers/flattening/arrays.cpp [hunk 24]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+  wegt::node_indext a=path.front().n;
+  wegt::node_indext b=path.back().n;
+
+  // do constraints
+  const index_sett &index_set_a=index_map[a];
+  const index_sett &index_set_b=index_map[b];
+
+  std::cout << "b is: "
+            << from_expr(ns, "", arrays[b])
+            << '\n';
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: don't write on std::cout

File: src/solvers/flattening/arrays.cpp [hunk 25]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+
+  for(const auto &index_a : index_set_a)
   {
-    // lazily add the constraint
-    if(incremental_cache)
+    // lazily instantiate read-over-write
+    if(arrays[b].id()==ID_with)
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: use expr_try_dynamic_cast

File: src/solvers/flattening/arrays.cpp [hunk 26]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+      // 2. add (i=j AND path_cond)=>a[i]=b[j]
+      // 3. The path condition requires the update indices
+      //    on the stack to be different from i.
+      exprt::operandst cond;
+      cond.reserve(path.size()+1);
+      cond.push_back(equal_exprt(index_a, index_b));
+      weg_path_condition(path, index_a, cond);
+
+      // a_i=b_i
+      index_exprt a_i(arrays[a], index_a);
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp:343]: const?

File: src/solvers/flattening/arrays.cpp [hunk 27]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+      // cond => a_i=b_i
+      implies_exprt implication(
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: const?

File: src/solvers/flattening/arrays.cpp [hunk 28]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+        conjunction(cond),
+        equal_exprt(a_i, value_b));
+      std::cout << "C2a: "
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: don't write on std::cout

File: src/solvers/flattening/arrays.cpp [hunk 29]
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
...
+                << from_expr(ns, "", implication)
+                << '\n';
+      set_to_true(implication);
     }
-    else
+    else if(arrays[b].id()==ID_array_of)
```
Comment (romainbrenguier) [inline src/solvers/flattening/arrays.cpp]: use expr_try_dynamic_cast

**Review by romainbrenguier — DISMISSED** [review]
I only partially reviewed, but several issues need to be addressed.

**Comment by martin-cs** [comment]
@tautschnig shout when you want a review from me.

**Comment by TGWDB** [comment]
@tautschnig This PR is still waiting on some changes to address reviewer comments (and now a rebase). Is there a plan to progress on this PR since it is mentioned in others?

**Comment by tautschnig** [comment]
> @tautschnig This PR is still waiting on some changes to address reviewer comments (and now a rebase). Is there a plan to progress on this PR since it is mentioned in others?

Yes, this is on my pile of TODOs. In terms of performance, it's probably one of the most urgent to work on. Just trying to address the soundness-related ones first.

**Comment by martin-cs** [comment]
@tautschnig : Trevor and I have been having some issues with the current array code and may need to look at it in the next few months.  Reviewing this / getting it merged might well help us with that.

**Comment by TGWDB** [comment]
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

**Pull request closed without merging**